### PR TITLE
fix(consensus): do not precommit nil if proposal block is received

### PR DESCRIPTION
**Pull request by cason**
Addresses #2119.

ADR 112, describing the PBTS implementation, [states](https://github.com/cometbft/cometbft/pull/2223/files#diff-81d55f6e52eabd6cce264a76fba015de46092b9ccda241de10cd265890ab8ad9R302):

    The precommit step will not require much modification.
    Its proposal validation rules will change in the same ways that validation will change in the prevote step with the exception of the `timely` check: precommit validation will never check that the timestamp is `timely`.

The changes proposed here are in line with this description.

Moreover, this PR in fact revert the changes to the precommit step (`enterPrecommit` method) added by https://github.com/tendermint/tendermint/pull/7480 and https://github.com/tendermint/tendermint/pull/7391.

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments
- [x] Title follows the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) spec

### CODE REVIEW

File: internal/consensus/state.go [hunk 1]
```diff
@@ -1615,20 +1615,14 @@ func (cs *State) enterPrecommit(height int64, round int32) {
 	}
 	// At this point, +2/3 prevoted for a particular block.
 
-	// If we never received a proposal for this block, we must precommit nil
-	if cs.Proposal == nil || cs.ProposalBlock == nil {
+	// If we did not receive a proposal block to vote for, we must precommit nil
+	// FIXME: although +2/3 prevoted for the block, therefore it is valid.
```
Comment (mzabaluev) [inline internal/consensus/state.go:1619]: Who's the FIXME label for? As stated in #2119, this is correct accordingly to the spec, what's there for the code monkeys to fix?

**Review by mzabaluev — APPROVED** [review]
Looks good, with a nitpick on a comment.

File: internal/consensus/state.go (hunk 1 above)
Comment (cason) [reply internal/consensus/state.go:1619]: Good point. I think the question here is that when adapting the code to match the spec we are losing some opportunities to, in a smart way, speed up decisions when we have Byzantine proposers, for example.

In our in production (current) releases, we can prevote and precommit to a block that matches our locked block. Because this is safe (at least, I don't see how it would not be safe). This code, for the precommit step is still here, btw, see lines 1633/1627.

When we require having `cs.ProposalBlock` to precommit a block, we are not using the possibility added to the above mentioned logic/lines.

In other words, what we would need here is a way to re-use the `LockedBlock`, setting `cs.Proposal` to it, in case the block we are looking for matches the locked one. This save us, in particular, the cost of downloading it again in a round `r > cs.LockedRound`. We may also consider the `cs.ValidRound`, for the same reasons. But this should be addressed by another issue/PR.

Comment (cason) [reply internal/consensus/state.go:1619]: If you navigate the history of commits, you realize that I removed this check altogether. The problem of which is that `TestStateLock_MissingProposalWhenPOLForLockedBlock` test fails. I then reintroduced this check to make this test to pass.

The fun part is that the description of the test does not match the test itself. Which is odd.

**Comment by lasarojc** [comment]
LGTM

File: internal/consensus/state.go (hunk 1 above)
Comment (mzabaluev) [reply internal/consensus/state.go]: Hmm, not sure, maybe the FIXME explanation should be expanded or placed elsewhere, to reflect on the missed opportunities and/or the oddly specified test, and the note should state what it does now.

**Review by sergio-mena — APPROVED** [review]
👍
